srml/slashing/src/lib.rs

```diff
@@ -0,0 +1,186 @@
+//!		fn severity(&self, k: u64, n: u64) -> Fraction<Self::Severity> {
+//!			let numerator = 20 * n;
+//!			let denominator = 3*k - 3;
+//!
+//!			if numerator / n >= 1 {
+//!				Fraction::new(1, 20)
+//!			} else {
+//!				Fraction::new(denominator, numerator)
+//!			}
+//!		}
+//!	}
+//!
+//! struct Balance<T>(PhantomData<T>);
+//!
+//! impl<T: Trait> OnSlashing<T::AccountId> for Balance<T> {
+//!
+//!		fn on_slash<M: Misconduct>(who: &T::AccountId, severity: Fraction<M::Severity>) {
+//!			// This doesn't compile, see `srml/staking/slash.rs` for a more elaborate example
+//!			// let balance = T::Currency::free_balance(who);
+//!			// let slash = balance / severity;
+//!			// T::Currency::on_slash(who, balance);
+//!		}
+//! }
+//!
+//!	struct SlashingWrapper<T>(PhantomData<T>);
+//!
+//!	impl<T: Trait> Slashing<T::AccountId> for SlashingWrapper<T> {
+//!		type Slash = Balance<T>;
+//!
+//!		fn slash<CM: ContinuousMisconduct>(who: &T::AccountId, misconduct: &mut CM) -> u8 {
+//!			let severity = misconduct.severity();
+//!			Self::Slash::on_slash::<CM>(&who, severity);
+//!			misconduct.on_misconduct();
+//!			severity.as_misconduct_level()
+//!		}
+//!
+//!		fn slash_on_checkpoint<EM: EraMisconduct>(
+//!			misbehaved: &[T::AccountId],
+//!			total_validators: u64,
+//!			misconduct: &EM
+//!		) -> u8 {
+//!			let severity = misconduct.severity(misbehaved.len() as u64, total_validators);
+//!
+//!			for who in misbehaved {
+//!				Self::Slash::on_slash::<EM>(who, severity);
+//!			}
+//!
+//!			severity.as_misconduct_level()
+//!		}
+//!	}
+//!
+//! fn main() {
+//!		let unresponsive = Unresponsive;
+//!		let misbehaved_validators = vec![1, 2, 3, 4, 5, 6];
+//!		let total_validators = 100;
+//!		// MyModule is type that implements `Trait`
+//!		// SlashingWrapper::<MyModule>::slash_on_checkpoint(&misbehaved_validators, total_validators, &unresponsive);
+//! }
+//! ```
+
+use parity_codec::Codec;
+use primitives::traits::{SimpleArithmetic, MaybeSerializeDebug};
+
+/// Pre-defined types
+pub mod misconduct;
+
+mod types;
+
+pub use types::Fraction;
+
+// The specification specifices four different misconduct levels:
+//		1) Slashing: 0 <= x <= 0.001
+//		2) Slashing: 0.001 < x <= 0.01
+//		3) Slashing: 0.01 < x <= 0.1
+//		4) Slashing: 0.1 < x <= 1.0
+type MisconductLevel = u8;
+
+/// Base trait for representing misconducts
+pub trait Misconduct {
+	/// Severity represented as a fraction
+	type Severity: SimpleArithmetic + Codec + Copy + MaybeSerializeDebug + Default + Into<u128>;
+}
+
+/// Behaviour based on misconduct reporting on end of era / timeslot.
+/// In those cases the reporter will keep state and report it.
+pub trait EraMisconduct: Misconduct {
+
+	/// Estimate severity based `number of misbehaved validators` and `number of validators`
+	// TODO(niklasad1): shall `num_misbehaved` & `num_validators` be generic?!
+	fn severity(&self, num_misbehaved: u64, num_validators: u64) -> Fraction<Self::Severity>;
```

Comment:
grumble: merge traits `EraMisconduct` and `ContinuousMisconduct` and have something like Rob suggested: `rolling window of k*era_len where k is probably 1 or 2`